### ISSUE TITLE
builtins:SystemExit error:1

### ISSUE THREAD

**Issue by sarah-clements** [mozilla/mozci#458]
The mozci integration is live on production - using version 2.0.2 - but I'm seeing a lot of new errors in New Relic and the heroku metrics that I think are causing timeouts. Mostly it's happening with the push health summary API, which is called in the treeherder jobs (for try only). I've seen this timeout happen for a certain push that then later succeeds and shows results.

This is part of the stack trace:
```
File "/app/.heroku/python/lib/python3.7/site-packages/rest_framework/views.py", line 506, in dispatch
File "/app/treeherder/webapp/api/push.py", line 335, in health_summary
File "/app/treeherder/webapp/api/push.py", line 44, in _get_failure_data
File "/app/.heroku/python/lib/python3.7/site-packages/mozci/push.py", line 956, in get_likely_regressions
File "/app/.heroku/python/lib/python3.7/site-packages/mozci/util/memoize.py", line 28, in method_call
File "/app/.heroku/python/lib/python3.7/site-packages/mozci/push.py", line 865, in get_regressions
File "/app/.heroku/python/lib/python3.7/site-packages/mozci/push.py", line 758, in get_candidate_regressions
File "/app/.heroku/python/lib/python3.7/site-packages/mozci/push.py", line 647, in _iterate_failures
File "/app/.heroku/python/lib/python3.7/site-packages/mozci/util/memoize.py", line 49, in __get__
File "/app/.heroku/python/lib/python3.7/site-packages/mozci/push.py", line 421, in label_summaries
File "/app/.heroku/python/lib/python3.7/site-packages/mozci/util/memoize.py", line 49, in __get__
File "/app/.heroku/python/lib/python3.7/site-packages/mozci/push.py", line 290, in tasks
File "/app/.heroku/python/lib/python3.7/concurrent/futures/_base.py", line 301, in wait
File "/app/.heroku/python/lib/python3.7/threading.py", line 552, in wait
File "/app/.heroku/python/lib/python3.7/threading.py", line 296, in wait
File "/app/.heroku/python/lib/python3.7/site-packages/gunicorn/workers/base.py", line 201, in handle_abort
```
Any ideas?

**Comment by sarah-clements** [comment]
I was looking at the metrics closer and my guess is this is causing too much of a web load to be querying mozci in every push that is loaded in the jobs view. I'm going to try adjusting that to see if it helps.

**Comment by marco-c** [comment]
It looks like it is timing out while retrieving groups for the tasks in the push (https://github.com/mozilla/mozci/blob/e488ccd4966163ed9532ab51f4584fc60286ae10/mozci/push.py#L280).
I wonder if this isn't because of the errorsummary.log fallback. @ahal wdyt if we assumed Treeherder data is complete and returned an empty list instead of raising a ContractNotFilled error?

**Comment by ahal** [comment]
We could do that, it just means that we won't be able to get this data on
older pushes. I guess that is fine.

Or the treeherder API could be adjusted to return a 404 when querying a
push with no data.

**Comment by ahal** [comment]
Yeah no worries. If not falling back to errorsummary here makes things
better let's just do that. I guess we don't have any use cases of consumers
using really old pushes anyway.

**Comment by ahal** [comment]
@sarah-clements Actually, I think the even better solution here would be to remove the `errorsummary` data source from your config. Since that data source only provides the test groups / errors, you won't lose any additional data coverage by doing so.

**Comment by sarah-clements** [comment]
> @sarah-clements Actually, I think the even better solution here would be to remove the `errorsummary` data source from your config. Since that data source only provides the test groups / errors, you won't lose any additional data coverage by doing so.

Do you mean the mozci config file? I've never included `errorsummary`. The push-health-mozci pr has `data_sources = ['treeherder_db', 'hgmo', 'taskcluster']`

**Comment by marco-c** [comment]
There is something fishy if you don't use the `errorsummary` data source, since `retrieve_results` should be super fast with `treeherder_db`...

**Comment by marco-c** [comment]
Maybe a deadlock caused by the lock at https://github.com/mozilla/mozci/blob/72b83507fd001c782f207b3093c7f35e5a7e68a0/mozci/data/sources/treeherder/__init__.py#L37 and the usage of concurrent.futures.wait?

**Comment by marco-c** [comment]
From https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor:
> Deadlocks can occur when the callable associated with a Future waits on the results of another Future.

This is not exactly our case because our callables are not waiting on the results of another Future directly, but they are indirectly (a callable is waiting on the lock held by another). So, not sure if it applies to us, but might be.

**Comment by ahal** [comment]
I don't think that would be it, since any thread that acquires that lock should release it before it returns its result.

But I guess we could try to remove that lock and see if it works anyway.. Unsure how we can test this without deploying on treeherder though. Maybe we could provide a patch and @sarah-clements could try applying it directly in her site-directories folder?

Another shot in the dark to fix this might be to use `asyncio` rather than `concurrent.futures`.

**Comment by sarah-clements** [comment]
> I don't think that would be it, since any thread that acquires that lock should release it before it returns its result.
> 
> But I guess we could try to remove that lock and see if it works anyway (it means we might have multiple calls to treeherder db for the same information).. Unsure how we can test this without deploying on treeherder though. Maybe we could provide a patch and @sarah-clements could try applying it directly in her site-directories folder?
> 
I could try testing a patch out locally. Do we know for certain it only happens with `treeherder_db`? Maybe I should deploy the push-health-mozci on prototype and use the `treeherder_client` as primary source instead to verify.

**Comment by ahal** [comment]
That would be helpful! Also are there timestamps in the log? Wondering how
long that timeout is and where it's coming from.

**Comment by sarah-clements** [comment]
> That would be helpful! Also are there timestamps in the log? Wondering how long that timeout is and where it's coming from.

I deployed the push-health-mozci-v2 (same as the original, just renamed) branch to prototype earlier today, and made sure it's utilizing `treeherder_client` as primary source.  I just checked New Relic and I can see the exact same errors/stack trace. It's originating primarily from the `health_summary` API, which is called in the jobs view for all try pushes (used to show the push health badges Cam implemented a while ago, and the "debug with push health" widget).
![Screen Shot 2021-05-18 at 1 40 55 PM](https://user-images.githubusercontent.com/19615783/118720679-c6270200-b7de-11eb-828e-35f9c08d93ff.png)

Side note: I'll probably remove that badge at some point once Push Health becomes the default view.

One of the errors I saw in New Relic is for push bd4ba6133657ed98377e76af60ea8fc3bb683f5e on try. I checked the [papertrail logs](https://my.papertrailapp.com/groups/1510504/events?q=mozci+bd4ba6133657ed98377e76af60ea8fc3bb683f5e) for prototype, compared the timestamp and saw that it seemed to stop at retrieving regressions at the time of failure, even though previously it seemed to retrieve it successfully (push health API is queried every few minutes until a push's jobs are complete):
`2021-05-18T22:43:40.000Z treeherder-prototype app/web.1 22:43:40.294 | DEBUG    |      mozci.push:756 - Retrieving candidate regressions for bd4ba6133657ed98377e76af60ea8fc3bb683f5e... 
`

And I saw this in [New Relic](https://one.nr/0gbRKPVZ9jE) - it looks like its getting hung up on HG requests.
![Screen Shot 2021-05-18 at 3 55 41 PM](https://user-images.githubusercontent.com/19615783/118733774-caa8e600-b7f1-11eb-8317-f5a51cbcc064.png)

Does this help?

**Comment by ahal** [comment]
I'm wondering if this might have been caused by issue #473. If the assertion wasn't getting logged, it may have caused timeouts. And I think the observation that it at first didn't work and then later worked also fits as that issue happened when tasks weren't `completed` yet. @sarah-clements could you try this again using mozci 2.0.3?

**Comment by sarah-clements** [comment]
> I'm wondering if this might have been caused by issue #473. If the assertion wasn't getting logged, it may have caused timeouts. And I think the observation that it at first didn't work and then later worked also fits as that issue happened when tasks weren't `completed` yet. @sarah-clements could you try this again using mozci 2.0.3?

Sorry, I missed this. I'll try it out end of this week or early next week.
